### ISSUE TITLE
not hasattr(...) causes conditional block to resolve self to Never

### ISSUE THREAD

**Issue by moschi** [astral-sh/ty#2455]
### Summary

The following code snipped illustrates the issue:

```python
class NeverWithHasAttr:
    @property
    def property_a(self) -> str:
        if not hasattr(self, "_property_a"):
            self._property_a = f"{self.this_does_not_exist} does not cause an error, self is Never"
        return self._property_a

    @property
    def property_b(self) -> str:
        if getattr(self, "_property_b", None) is None:
            self._property_b = f"{self.this_does_not_exist} causes an error"
        return self._property_b

```

where `self` inside the `not hasattr(self, "_property_a"):` block is Never, thus not causing any misspelled properties to be flagged as errors.

### Version

ty 0.0.11

**Comment by carljm** [comment]
Thanks for the report!

Given that we don't (currently) distinguish between maybe-bound and definitely-bound attributes, I guess we can't safely do negative narrowing on `hasattr`, only positive.
